control-plane/agents/core/tests/volume/mod.rs

```diff
@@ -1916,3 +1916,264 @@ async fn smoke_test(cluster: &Cluster) {
+}
+
+// Given: 2 Nodes, 2 Pools, 2 Volumes with 2 replicas each, 1 Volume published on some node.
+// Scenario: When the second volume is to be published it should not choose the same node
+// to ensure equal distribution of targets.
+async fn target_distribution(cluster: &Cluster) {
+    let volume_client = cluster.grpc_client().volume();
+    // Publish volume1
+    let pub_vol1 = volume_client
+        .publish(
+            &PublishVolume {
+                uuid: VolumeId::try_from(VOLUME_1).unwrap(),
+                target_node: None,
+                ..Default::default()
+            },
+            None,
+        )
+        .await
+        .expect("Should not fail");
+
+    // Publish volume2
+    let pub_vol2 = volume_client
+        .publish(
+            &PublishVolume {
+                uuid: VolumeId::try_from(VOLUME_2).unwrap(),
+                target_node: None,
+                ..Default::default()
+            },
+            None,
+        )
+        .await
+        .expect("Should not fail");
+
+    // Both volume's target node should be different
+    assert_ne!(
+        pub_vol1.state().target.unwrap().node,
+        pub_vol2.state().target.unwrap().node
+    );
+
+    // Cleanup
+    let _ = volume_client
+        .unpublish(
+            &UnpublishVolume::new(&VolumeId::try_from(VOLUME_1).unwrap(), false),
+            None,
+        )
+        .await
+        .expect("The volume should be unpublished");
+    let _ = volume_client
+        .unpublish(
+            &UnpublishVolume::new(&VolumeId::try_from(VOLUME_2).unwrap(), false),
+            None,
+        )
+        .await
+        .expect("The volume should be unpublished");
+}
+
+// Given: 2 Nodes, 2 Pools, 2 Volumes with 2 replicas each, 1 Volume published on some node.
+// Scenario 1: The node2 is killed, when the second volume is to be published it should choose
+// the same node for publish as there is no other node.
+// Scenario 2: The node1 and node2 both killed, when the second volume is to be published it should
+// fail with ResourceExhausted error.
+async fn offline_node(cluster: &Cluster) {
+    let volume_client = cluster.grpc_client().volume();
+
+    // Publish volume1
+    let pub_vol1 = volume_client
+        .publish(
+            &PublishVolume {
+                uuid: VolumeId::try_from(VOLUME_1).unwrap(),
+                target_node: None,
+                ..Default::default()
+            },
+            None,
+        )
+        .await
+        .expect("Should not fail");
+
+    // determine in which node the volume1 was published
+    let kill_node = if pub_vol1.state().target.unwrap().node.as_str() == "io-engine-1" {
+        "io-engine-2"
+    } else {
+        "io-engine-1"
+    };
+
+    // kill the node where the volume was published
+    cluster
+        .composer()
+        .stop(kill_node)
+        .await
+        .unwrap_or_else(|_| panic!("The {} container should stop", kill_node));
+
+    sleep(Duration::from_secs(4)).await;
```

Comment:
Do we have to wait so long here?

Reply:
Well I added `4s` as the node takes `2s` to change the state from Online, so 4 to be on safer side.

Reply:
But we're stopping io-engine gracefully with stop, so the core agent will change the state when it receives the stop message?

Reply:
Ahh right. It will deregister. Now that you have pointed, should it have been a `kill` here?

Reply:
I don't think it should make any difference for what we're testing for here?

Reply:
We are testing, if a node was previously being not selected due to equal distribution, will now be selected if there are no other nodes to select.

Reply:
but why does `stop` or `kill` make a difference here?

Reply:
Difference if graceful shutdown and abrupt shutdown. But, yeah checking for any of these two should suffice.

Reply:
Done.